test_util/integration_test.py

```diff
@@ -1422,3 +1424,159 @@ def test_mesos_agent_role_assignment(cluster):
     for agent in cluster.slaves:
         r = requests.get('http://{}:5051/state.json'.format(agent))
         assert r.json()['flags']['default_role'] == '*'
+
+
+def _get_snapshot_list(cluster):
+    list_url = '/system/health/v1/report/snapshot/list/all'
+    response = cluster.get(path=list_url).json()
+    logging.info('GET {}, response: {}'.format(list_url, response))
+
+    snapshots = []
+    for _, snapshot_list in response.items():
+        if snapshot_list is not None and isinstance(snapshot_list, list) and len(snapshot_list) > 0:
+            # append snapshots and get just the filename.
+            snapshots += map(lambda s: os.path.basename(s), snapshot_list)
+    return snapshots
+
+
+def test_3dt_snapshot_create(cluster):
+    """
+    test snapshot create functionality
+    """
+
+    # start the snapshot job
+    create_url = '/system/health/v1/report/snapshot/create'
+    response = cluster.post(path=create_url, payload={"nodes": ["all"]}).json()
+    logging.info('POST {}, response: {}'.format(create_url, response))
+
+    # make sure the job is done, timeout is 5 sec, wait between retying is 1 sec
+    status_url = '/system/health/v1/report/snapshot/status/all'
+
+    @retrying.retry(stop_max_delay=5000, wait_fixed=1000)
+    def wait_for_job():
+        response = cluster.get(path=status_url).json()
+        logging.info('GET {}, response: {}'.format(status_url, response))
+
+        # check `is_running` attribute for each host. All of them must be False
+        for _, attributes in response.items():
+            assert not attributes['is_running']
+
+    wait_for_job()
+
+    # the job should be complete at this point.
+    # check the listing for a zip file
+    snapshots = _get_snapshot_list(cluster)
+    assert len(snapshots) == 1, 'snapshot file not found'
+
+
+def test_3dt_snapshot_download_and_extract(cluster):
+    """
+    test snapshot download and validate zip file
+    """
+
+    snapshots = _get_snapshot_list(cluster)
+    assert snapshots
+
+    expected_common_files = ['dmesg-0.output.gz', 'opt/mesosphere/active.buildinfo.full.json.gz', '3dt-health.json']
+
+    # these files are expected to be in archive for a master host
+    expected_master_files = ['dcos-mesos-master.service.gz'] + expected_common_files
+
+    # for agent host
+    expected_agent_files = ['dcos-mesos-slave.service.gz'] + expected_common_files
+
+    # for public agent host
+    expected_public_agent_files = ['dcos-mesos-slave-public.service.gz'] + expected_common_files
+
+    try:
+        tmp_dir = tempfile.mkdtemp()
```

Comment:
Python has https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory to make it so you don't have to hand write the cleanup it in the try / except here.